src/Formatting/Combinators.hs

```diff
@@ -0,0 +1,515 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-|
+Module      : Formatting.Combinators
+Copyright   : (c) 2020 Alex Chapman
+License     : BSD3
+Maintainer  : alex@farfromthere.net
+Stability   : experimental
+Portability : GHC
+Description : Formatting combinators for building new formatters, with some useful pre-defined formatters.
+
+A formatting combinator takes a Format and returns another Format.
+Generally we want to change what the original format takes as its *input*,
+leaving the output polymorphic.
+Many of these combinators can be chained together to form a single 'Format'.
+
+Implementation detail: in order to be able to chain multiple combinators to make a single 'Format' we need them all to use the same intermediate string type, and we have chosen 'Builder'.
+This does not tie you to using 'Builder's, because the final output string type 'r' is still polymorphic.
+|-}
+module Formatting.Combinators
+  (
+  -- * Formatting lists of data
+    concatenated
+  , joinedWith
+  , intercalated
+  , unworded
+  , unlined
+  , spaced
+  , commaSep
+  , commaSpaceSep
+  , list
+  , took
+
+  -- * Splitting strings to pass to other formatters
+  , splat
+  , splatWith
+  , splatOn
+  , worded
+  , lined
+
+  -- * Altering formatted strings
+  , alteredWith
+  , replaced
+  , uppercased
+  , lowercased
+  , titlecased
+  , truncated
+  , midTruncated
+
+  -- * Wrapping formatted strings
+  , prefixed
+  , suffixed
+  , surrounded
+  , enclosed
+  , quoted
+  , doubleQuoted
+  , parenthesised
+  , squareBracketed
+  , curlyBracketed
+  , angleBracketed
+  , backticked
+
+  -- * Changing indentation
+  , indented
+  , indentedLines
+  , reindented
+
+  -- * Numerical adapters
+  , roundedTo
+  , truncatedTo
+  , ceilingedTo
+  , flooredTo
+
+  -- * Lens formatting
+  , viewed
+  ) where
+
+import Control.Category ((>>>))
+import Data.Foldable (toList)
+import Data.Function ((&))
+import Data.Functor.Const (Const(..), getConst)
+import Data.Int (Int64)
+import Data.Text.Lazy (Text)
+import qualified Data.Text.Lazy as TL
+import Data.Text.Lazy.Builder (Builder)
+import qualified Data.Text.Lazy.Builder as TLB
+import Formatting.Internal
+import Formatting.Formatters
+
+-- | Format each value in a list and concatenate them all:
+--
+-- >>> format (concatenated text) ["one", "two", "three"]
+-- "onetwothree"
+--
+-- >>> format (took 15 (concatenated bin)) [1..]
+-- "1101110010111011110001001101010111100110111101111"
+concatenated :: Foldable t => Format Builder (a -> Builder) -> Format r (t a -> r)
+concatenated f = later $ toList
```

Comment:
Is there a simpler expression using `foldMap`?

Reply:
Yes! I'll update that.